### PR TITLE
feat: add support for bsf update for non-semver packages

### LINKED ISSUE #29: bsf update for non-semver packages
We've got bsf update working for semver. From our database, we noticed ~80% of packages follow semver, only 97 follow strict calver and rest are considered "unknown". Opening this issue to track as 20% is significant amount.

I think following things need to be done:

Deeper analysis to see if various conventions of calver are used. Gut feel is there should be more calver conventions packages are following other than YYYY.MM.x.

Discuss with community if they would be happy if we update to latest package when a constraint is given and it's not semver or calver, typically indicating revision numbers, commit hashes,etc.

### PR DESCRIPTION
**Pull request by hanshal101**
Closes #29 
Fixes #29 

## Description
This PR adds supports for `non-semver` packages in `bsf update` command.

### CODE REVIEW

File: cmd/update/update.go [hunk 1]
```diff
@@ -123,6 +123,12 @@ func parsePackagesForUpdates(versionMap map[string]*buildsafev1.FetchPackagesRes
 		updateType := update.ParseUpdateType(k)
 
 		switch updateType {
+		case update.UpdateTypeDate:
```
Comment (dr-housemd) [inline cmd/update/update.go:126]: This should be mentioned on L31 for the help menu.

File: pkg/update/update.go [hunk 2]
```diff
@@ -42,6 +46,28 @@ func ParsePackage(pkg string) (name, version string) {
...
+func GetDateBasedVersion(v *buildsafev1.FetchPackagesResponse, version string) string {
+	if v == nil {
+		return ""
+	}
+
+	var latestVersion string
+	var latestDate string
+
+	for _, pkg := range v.Packages {
+		if strings.HasPrefix(pkg.Version, version[:4]) {
```
Comment (dr-housemd) [inline pkg/update/update.go:59]: There is a pkg.EpochSeconds field . We need to use that instead of `pkg.Version`.

Comment (hanshal101) [reply pkg/update/update.go:59]: something like this `string(pkg.EpochSeconds)` ?

File: cmd/update/update.go [hunk 3]
```diff
@@ -28,7 +28,7 @@ var UpdateCmd = &cobra.Command{
 	Long: `Updates can be done for development and runtime dependencies based on constraints. Following constraints are supported:
 		~ : latest patch version
 		^ : latest minor version
-
+		# : latest calver version
```
Comment (dr-housemd) [inline cmd/update/update.go:31]: it's not calver. We can just say "Date based updates"

File: cmd/update/update.go [hunk 4]
```diff
@@ -28,7 +28,7 @@ var UpdateCmd = &cobra.Command{
 	Long: `Updates can be done for development and runtime dependencies based on constraints. Following constraints are supported:
 		~ : latest patch version
 		^ : latest minor version
-
+		# : date based updates
```
Comment (RealHarshThakur) [inline cmd/update/update.go:31]:
```suggestion
		# : date based updates. Useful, when a package doesn't follow semver and we want to update to the latest published version
```

File: pkg/update/update.go [hunk 5]
```diff
@@ -42,6 +47,20 @@ func ParsePackage(pkg string) (name, version string) {
 	return nameWithVersion[0], version
 }
 
+// GetDateBasedVersion returns the latest date version for the given version.
+func GetDateBasedVersion(v *buildsafev1.FetchPackagesResponse, version string) string {
```
Comment (RealHarshThakur) [inline pkg/update/update.go:51]: What's the `version` argument for? Is it needed or can be dropped?

Comment (dr-housemd) [inline pkg/update/update.go:51]: Is the version arg required here?

Comment (hanshal101) [reply pkg/update/update.go:51]: nope

**Comment by hanshal101** [comment]
All changes CLEARED !!!

**Review by dr-housemd — APPROVED** [review]

**Pull request merged**